### ISSUE TITLE
Oversized name columns found in steps table while migrating from sqlite to postgres

### ISSUE THREAD

**Issue by gerard** [buildbot/buildbot#4599]
I'm currently migrating a buildbot sqlite database to postgres.  The initial database was created around 0.9.11.  It currently runs buildbot 1.5.0 and will run 2.0 after the migration.

While converting the database I've found out that we have been using step names over 50 chars, when the model should not allow more than that.  Sqlite inserted the oversized columns without issue (no truncation!) but postgres is now nagging that it can't INSERT that data.

In order to proceed, I've simply patched the schema in the sqlite dump to use a VARCHAR(256) and I've modified model.py to match.

This is obviously a bad solution and I would prefer not to have to patch our buildbot master to have it running in our database.  Other than just truncating the fields, is there anything that can be done to fix the migration?

**Comment by rjarry** [comment]
Hi @gerard,

Unfortunately, SQLite does not properly enforce a lot of constraints... I do not see any better solution other than truncating.

However, maybe we could force the truncation in buildbot code to avoid these problems in the future.

**Comment by gerard** [comment]
Truncating or failing ahead of time sounds like a good idea.  It could be a setting where you can choose to ignore it, maybe SQLAlchemy could have something for that?

I think most people using buildbot will simply start with sqlite and then move to other databases eventually.  Having to face this issue in migration is a bit frustrating, since losing data is not nice. Besides, (buildid, name) is supposed to be unique according to the schema, so the manual truncation is not trivial.

That said, I have 2 slightly offtopic comments:
 * 50 chars for the step name seems a bit on the short side :P.
 * Buildbot documentation is great, and partly the reason we are using it.  However, there's nothing on how to do a db migration (or I haven't found it?).  You could argue SQLAlchemy should take care of everything, but pitfalls like these should be documented.  I'm slightly worried that there could be other issues that I might have missed.

**Comment by tardyp** [comment]
There are so much combinaisons that it is difficult for me to have one receipe for a data migration between DB.

I did handle such several times in prod, and what we did it to just create a shiny new master, and keep the old one with the old db for reference aside.

**Comment by rjarry** [comment]
@gerard wrote:
> I think most people using buildbot will simply start with sqlite and then move to other databases eventually. Having to face this issue in migration is a bit frustrating, since losing data is not nice. Besides, (buildid, name) is supposed to be unique according to the schema, so the manual truncation is not trivial.

I think we should be more explicit about what DB engine to use and when. If you indent on using buildbot in production, SQLite is *NOT* a good solution. The documentation does not recommend anything as far as I know.

> * 50 chars for the step name seems a bit on the short side :P.

Indeed, maybe we could increase this. There are not that much steps and the disk usage should not be affected.

> * Buildbot documentation is great, and partly the reason we are using it.  However, there's nothing on how to do a db migration (or I haven't found it?).  You could argue SQLAlchemy should take care of everything, but pitfalls like these should be documented.  I'm slightly worried that there could be other issues that I might have missed.

I agree with @tardyp, if you need to switch DB engines, it may be better to simply drop all build history and start from scratch with a new database. Maybe that's something we could talk about in the documentation.

**Comment by gerard** [comment]
I would agree that simply deploying a new instance on a new host is a fine solution most of the time.  That's what we did when we upgraded to 0.9 and it was fine, but doing that now would require more work and wouldn't be transparent to the users, so it has its drawbacks.  In retrospect, our biggest mistake was not moving to postgres before.

> There are so much combinaisons that it is difficult for me to have one receipe for a data migration between DB.

Most users will be migrating to MySQL/PostgreSQL from sqlite (due to defaults).  Maybe some other DBs are supported, but only those 3 are mentioned on the documentation.  Providing migration scripts is indeed excessive and unnecessary (there are tools to do it, I used sequel but pgloader also exists).  On the other hand it would be nice to have any schema specific issues documented and looked at.

In any case, I will just go ahead with what we have: a one liner patch to the column size in model.py.  Hopefully you consider increasing it in the future so I can drop it ;)

Feel free to close the ticket.

**Comment by rjarry** [comment]
@gerard, if you have time to submit a pull request with a migration script + test, I'll be happy to review and merge it.

**Comment by gerard** [comment]
Ideally there should be no need to have a migration script.  This is possible if sqlalchemy or buildbot enforce the VARCHAR(size) constraints.

The whole process for me was:
 * sed the sqlite dump to fix VARCHAR()
 * Use sequel
 * Fix a few default values (ALTER COLUMN X SET DEFAULT) that sequel missed in postgres (I guess I should have mentioned this... I kinda forgot, but this to me is a sequel bug).

I don't mind sending a pull request with troubleshooting docs if you like.

**Comment by gerard** [comment]
I could also have a look at how to enforce the VARCHAR constraint on buildbot side.

**Comment by rjarry** [comment]
I meant a migration script to increase the size of the step name column.

**Comment by gerard** [comment]
> I meant a migration script to increase the size of the step name column.

Oh, right.  I'll try to have a look at it this weekend.

**Comment by YngveNPettersen** [comment]
We just hit the 50 char step name limit, too.

The stepname in question is constructed build time as "general name" + "filename to be up loaded", which in this case was ~60 characters.

I did a "ALTER TABLE steps ALTER name TYPE varchar(150);" in the DB to work around the issue.

**Comment by amvasilyev** [comment]
We had the similar issue with the MariaDB setup as a backend. The correct alteration request for the MariaDB is:
```sql
ALTER TABLE steps MODIFY name varchar(150);
```
